client/src/rr/dnssec/hash.rs

```diff
@@ -0,0 +1,370 @@
+               sig.sig_expiration(),
+               sig.sig_inception(),
+               sig.key_tag(),
+               sig.signer_name(),
+               records)
+}
+
+
+/// [RFC 4035](https://tools.ietf.org/html/rfc4035), DNSSEC Protocol Modifications, March 2005
+///
+/// ```text
+///
+/// 5.3.2.  Reconstructing the Signed Data
+///             ...
+///             To calculate the name:
+///                let rrsig_labels = the value of the RRSIG Labels field
+///
+///                let fqdn = RRset's fully qualified domain name in
+///                                canonical form
+///
+///                let fqdn_labels = Label count of the fqdn above.
+///
+///                if rrsig_labels = fqdn_labels,
+///                    name = fqdn
+///
+///                if rrsig_labels < fqdn_labels,
+///                   name = "*." | the rightmost rrsig_label labels of the
+///                                 fqdn
+///
+///                if rrsig_labels > fqdn_labels
+///                   the RRSIG RR did not pass the necessary validation
+///                   checks and MUST NOT be used to authenticate this
+///                   RRset.
+///
+///    The canonical forms for names and RRsets are defined in [RFC4034].
+/// ```
+pub fn determine_name(name: &Name, num_labels: u8) -> Option<Name> {
+    //             To calculate the name:
+    //                let rrsig_labels = the value of the RRSIG Labels field
+    //
+    //                let fqdn = RRset's fully qualified domain name in
+    //                                canonical form
+    //
+    //                let fqdn_labels = Label count of the fqdn above.
+    let fqdn_labels = name.num_labels();
+    //                if rrsig_labels = fqdn_labels,
+    //                    name = fqdn
+
+    if fqdn_labels == num_labels {
+        return Some(name.clone());
+    }
+    //                if rrsig_labels < fqdn_labels,
+    //                   name = "*." | the rightmost rrsig_label labels of the
+    //                                 fqdn
+    if num_labels < fqdn_labels {
+        let mut star_name: Name = Name::new().label("*");
+        let rightmost = name.trim_to(num_labels as usize);
+        if !rightmost.is_root() {
+            star_name.append(&rightmost);
+            return Some(star_name);
+        }
+        return Some(star_name);
+    }
+    //
+    //                if rrsig_labels > fqdn_labels
+    //                   the RRSIG RR did not pass the necessary validation
+    //                   checks and MUST NOT be used to authenticate this
+    //                   RRset.
+    // TODO: this should be an error
+    None
+}
+
+#[cfg(test)]
+mod tests {
+    extern crate openssl;
+    use self::openssl::rsa::Rsa;
+
+    use rr::{Name, RecordType};
+    use rr::rdata::SIG;
+    use op::{Message, Query, UpdateMessage};
+
+    pub use super::*;
+
+    #[test]
+    fn test_hash_rrset() {
+        let rsa = Rsa::generate(512).unwrap();
```

Comment:
Yes. I'll move all tests to start using checked in keys for tests, though probably not in this PR. I see the sanity in doing that, but I don't have a big motivation at the moment. It would remove a dependency on openssl for running tests, but no one has mentioned needing that yet.